Micro player

reduckted opened the pull request:
Implementation of #701. 

# 📔 Overview 

This is a "micro" player that provides similar functionality to the mini player with a few key differences.

* This micro player is displayed in a separate window. Unlike the mini player, you don't need to close the micro player to get back to the main window.
* The micro player has a much smaller height, so it's suitable for placing at the top or bottom of the screen without getting in the road too much.
* The track information is always displayed.
* The _Thumbs up_ and _Thumbs down_ buttons are always displayed.


# 🎨 Screenshots 

Turn it on or off from the side menu:
![image](https://user-images.githubusercontent.com/10321525/70128503-aeca8680-16c8-11ea-9e15-95b0431c1fe8.png)

On Windows (see limitations below):
![micro-player](https://user-images.githubusercontent.com/10321525/68464735-d7439a00-025c-11ea-8deb-d981a329dfda.gif)

On Linux:
![micro-player-linux](https://user-images.githubusercontent.com/10321525/68466333-e841da80-025f-11ea-8cbc-dcfee4651b60.gif)

With custom styles:
![dark-theme](https://user-images.githubusercontent.com/10321525/68465062-736da100-025d-11ea-9eda-a691611cbfd5.png)

# 🔧 Application changes

To allow the micro player window to know about what's happening in the main player, I've had to make some changes to the rest of the application.

## Playback API

The data in the playback API now has a `paused` property. I needed to add this to allow the micro player to know whether the player is stopped, paused or playing when the micro player first opens.

❓ Does this require an increase to the API version? It's This is an _addition_ to the public interface, so I'm assuming that it's _not_ considered a breaking change.

## Remove listeners from Emitter

There's an adapter class that forwards events from the `Emitter` class through to the micro player window. When the micro player closes, the event listeners need to be removed. I've added an `Emitter.removeListener()` function that just calls through to the underlying `ipcMain` object.

## "Application loaded" event

The micro player can open before the main window has finished loading GPM. I've made the `PlayerPage` emit an `"app:loaded"` event so that the micro player can show the loading indicator while the main window is still loading.

# 🔨 Testing

I developed it on Windows 10, so it's had plenty of testing on Windows, and I've poked it a _little_ bit on Linux - enough to find a few issues with it, but not enough to be super-confident with it. It has not been tested on macOS.

# ⚡️ Limitations

On Windows, it cannot be resized smaller than 39px high. This is due to a bug/limitation in Electron v3. This is fixed in Electron v6+.

# ☑️ To Do List

- [ ] Testing on Linux
- [ ] Testing on macOS
- [x] Fix bug with event listeners not being removed.

# What's next?

I've created this PR as a draft so that I can get some feedback on it. I haven't used Electron or React before, so please let me know if there's anything that I've done wrong on that side of things. 😁 

As I mentioned above, I've only really tested this on Windows with just some simple testing on Linux inside a VM, so it would be greatly appreciated if some people running Linux and macOS could install/build this and test it out.

Let me know what you think! ✌️

jostrander commented:
This is _awesome_. I looked through most everything and it looks good. The translations are properly filled in, the unit tests looks to be spot on, and the overall code quality is :fire:. CodeClimate has some complaints but not sure how important it is to please. 

A few comments based on initial testing:
1. It doesn't appear to respect the custom theme from the rest of the player.  Not a deal breaker, could be merged without, IMO.
2. It appears to work when switching to YTM, but the transition is kind of weird due to the micro player UI not resetting to match the new state of the player. 
3. The album art appears to be broken in YTM, but not sure that's related to this though. 
4. In Ubuntu (and maybe windows?) if you maximize using keyboard shortcut or by "docking" to the top of the screen it fills the screen. Not sure if electron allows you to disable this. 

The changes you made to the program seem to be relatively light, so I can't see them being much of an issue personally. 

So far everything looks great though, thanks for the quality PR! :tada:

reduckted commented:
> 1. It doesn't appear to respect the custom theme from the rest of the player. Not a deal breaker, could be merged without, IMO.

I've wrapped all of the styles for the micro player with a `[micro]` selector because I didn't want any new styles I added to break anything in the main player. :grimacing:

CSS specificity is the likely cause if your custom styles aren't flowing through to the micro player. 

I also thought it might be a good idea to allow the main app and the micro player to be styled independently so that you can, for example, use the default theme for the main window, but easily define a dark theme for the micro player. That can probably still be achieved without wrapping everything in a `[micro]` selector. I'm happy to investigate to see if there's some common class names I can use to allow the theme from the main window to flow into the micro player. 👍 

I'll post the CSS I've used for a dark theme below.

> 2. It appears to work when switching to YTM, but the transition is kind of weird due to the micro player UI not resetting to match the new state of the player.

Ah, I never tried it with YTM. When you say the transition is kind of weird, do you mean how the main window goes back to showing the loading indicator, but the micro player keeps showing the current track? I _think_ that if the application fired an event when switching modes, the micro player could listen for that event and revert back to its "loading" state. I'll have a play around and see if that would improve things.

> 3. The album art appears to be broken in YTM, but not sure that's related to this though.

I've just seen the same thing when I tried YTM. This seems to only be for some tracks though. The first track I tried didn't show any album art. I checked the image URL coming through from the playback API and it was https://music.youtube.com. The next track I tried had album art showing correctly, as did the one after that. 

It looks like there isn't a mini player in YTM mode (or at least I couldn't work out how to open it :grin:), so perhaps this is something that hasn't been fully implemented in the playback API for YTM yet.

> 4. In Ubuntu (and maybe windows?) if you maximize using keyboard shortcut or by "docking" to the top of the screen it fills the screen.

Oh, that's nasty. It doesn't happen on Windows, but I can replicate what you're seeing on my Ubuntu VM. I'll do some poking around and see if there's some way to stop that from happening.

> 5. Dragging only appears to work from the song information side of the micro player. I can't drag from the whitespace behind the controls themselves. Is this intended behavior?

Yeah, that's by design. My reasoning is that if you try to click a button and missed, you could potentially end up dragging the window, which wouldn't be very nice. Admittedly, I never tried putting the draggable area behind the buttons, so I can't say that it's what would actually happen though. I'll try it out with the drag area over the whole micro player. 👍 

Here's a sample dark theme that styles all of the elements in the micro player:

```css
[micro] {
  /* Use the default border color. */
  /* border-color: #ff5722 */
  background: #333333;
}

[micro] .info,
[micro] .controls button {
  fill: #cccccc;
  color: #cccccc;
}

[micro] .info .lg .artist-album,
[micro] .info .loader-container .loading-label {
  color: #959595;
}

[micro] .controls {
  background: #1e1e1e;
}

[micro] .controls button.play-pause {
  /* Use the default color. */
  /* color: #ff5722 */
}

[micro] .controls button:not([disabled]):active {
  background: #484848;
}

[micro] .controls button[disabled] {
  color: #545454;
  fill: #545454;
}
```

reduckted commented:
I've made some more changes and refactorings.

> CodeClimate has some complaints but not sure how important it is to please.

I've fixed as many of these as I can, mainly by refactoring code, but all of the "duplication" issues can't be fixed. I refactored some code to prevent duplication, but now it complains about the callers to the refactored functions. 🤷‍♂ 

> 1. It doesn't appear to respect the custom theme from the rest of the player. Not a deal breaker, could be merged without, IMO.

I was wrong about this being caused by CSS specificity 🤦‍♂. The markup in the micro player will only use the "Main App" styles, not the "Google Play Music" styles, and based on the two custom themes [listed in the wiki](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/wiki/Themes), they don't really style the main application that much.

> 2. It appears to work when switching to YTM, but the transition is kind of weird due to the micro player UI not resetting to match the new state of the player.

I've fixed this by adding an `"app:loading"` event that the main window fires when it first loads. Since switching between GPM and YTM will reload the main window, that event will fire and the micro player switches back into "loading" mode.

> 3. The album art appears to be broken in YTM, but not sure that's related to this though.

I've logged this issue in the `ytmusic.js` repo: https://github.com/gmusic-utils/ytmusic.js/issues/13

As a workaround, I've changed the micro player to use the placeholder album art if the album art URL starts with the YouTube Music URL.

> 4. In Ubuntu (and maybe windows?) if you maximize using keyboard shortcut or by "docking" to the top of the screen it fills the screen. Not sure if electron allows you to disable this.

I couldn't prevent this from happening 😞. I'm fairly confident that it's a bug in electron: https://github.com/electron/electron/issues/21595

> 5. Dragging only appears to work from the song information side of the micro player. I can't drag from the whitespace behind the controls themselves.

I tried making the drag handle cover the entire window, but unfortunately, the drag handle covering the same area as the buttons prevented the buttons from being clicked, even when the drag handle was below the buttons in the z-index.

I've also added a button in YTM mode to toggle the micro player on and off. It sits next to the "switch to GPM" button.

reduckted commented:
Oh, also the thumbs up, thumbs down and play/pause button currently don't show the correct state due to this bug in `gmusic.js`: https://github.com/gmusic-utils/gmusic.js/issues/26

The buttons still function correctly, they just don't reflect the current tracks' rating or play/paused state.

reduckted commented:
> It doesn't appear to respect the custom theme from the rest of the player. Not a deal breaker, could be merged without, IMO.

Massive facepalm moment. I didn't realize a custom theme could be applied. I thought the only way to change the appearance was via custom stylesheets. 🤦‍♂ Anyway, I've sorted that out and the custom theme is now applied to the micro player.

Default:
![image](https://user-images.githubusercontent.com/10321525/71552778-a60e6b80-2a4e-11ea-9c91-7615ab7d4455.png)

Light:
![image](https://user-images.githubusercontent.com/10321525/71552781-b45c8780-2a4e-11ea-9ce4-0e64bdac3df1.png)

Dark:
![image](https://user-images.githubusercontent.com/10321525/71552784-c0e0e000-2a4e-11ea-8222-674142a4a187.png)

And, of course, you can still override whatever bits you want using custom stylesheets, though it's changed from what I posted earlier.

```css
#micro-player .micro-player {
  border-color: red !important;
  background: orange;
}

#micro-player .info,
#micro-player .controls button,
#micro-player .light .info,
#micro-player .light .controls button,
#micro-player .dark .info,
#micro-player .dark .controls button {
  fill: white;
  color: white;
}

#micro-player .info .lg .artist-album,
#micro-player .info .loader-container .loading-label,
#micro-player .light .info .lg .artist-album,
#micro-player .light .info .loader-container .loading-label,
#micro-player .dark .info .lg .artist-album,
#micro-player .dark .info .loader-container .loading-label {
  color: green;
}

#micro-player .controls,
#micro-player .light .controls,
#micro-player .dark .controls {
  background: blue;
}

#micro-player .controls button.play-pause,
#micro-player .light .controls button.play-pause,
#micro-player .dark .controls button.play-pause {
  color: limegreen;
}

#micro-player .controls button:not([disabled]):active,
#micro-player .light .controls button:not([disabled]):active,
#micro-player .dark .controls button:not([disabled]):active {
  background: violet;
}

#micro-player .controls button[disabled],
#micro-player .light .controls button[disabled],
#micro-player .dark .controls button[disabled] {
  color: deeppink;
  fill: deeppink;
}
```
![image](https://user-images.githubusercontent.com/10321525/71552818-701db700-2a4f-11ea-9115-144e3ee69898.png)

Dragon911XX commented:
I absolutely love this. This is exactly what I was looking for when I made the request https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/issues/3770

A suggestion I would have, would maybe allowing the option to show/hide the extra options like ratings, play/pause, next/previous song, etc. Maybe also allowing us to pin it to the most front layer so no other windows are displayed on top. 

Keep on going with this, its amazing work!

reduckted commented:
@Dragon911XX Apologies for the very slow reply.

> A suggestion I would have...

They all sound like great ideas! I'll try to implement the ability to switch on and off the controls that you want to show.

> Also allowing us to pin it to the most front layer so no other windows are displayed on top.

The micro player already does that and, unlike the mini player, it can't be switched off. There are some limitations though, but I don't think we can get around them. On Windows, if you put the micro player over the taskbar it will stay above the taskbar. But if you focus the taskbar (for example, opening the start menu), then the taskbar will come forward over the top of the micro player. The same thing happens with the current mini player when it's configured to be always on top.

> Also, I'm not too savvy with github, but I'd LOVE to try this out. 

You can download windows installer [from the AppVeyor build result](https://ci.appveyor.com/project/MarshallOfSound/google-play-music-desktop-player-unofficial/builds/29795720/artifacts). Be aware that it won't contain any of the changes from the latest v4.7.1 release.

reduckted commented:
Hi @Dragon911XX, I've just pushed up the changes that allow you to control which buttons are visible in the micro player. There's now a tab in the desktop settings where you can choose what buttons you want to see.

![image](https://user-images.githubusercontent.com/10321525/77840973-5156ca00-71d0-11ea-8aca-19522164e52f.png)

I've also rebased onto the master branch, so this build will contain all of the changes from v4.7.1 as well. Here's the link: 
https://ci.appveyor.com/api/buildjobs/9ox4niylmim8f7x9/artifacts/dist%2Finstallers%2Fwin32%2FGoogle%20Play%20Music%20Desktop%20PlayerSetup.exe

reduckted commented:
I'm closing this since I don't use GPM or YouTube Music anymore, but if anyone is still interested in this, feel free to fork it.
